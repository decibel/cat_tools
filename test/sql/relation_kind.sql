\set ECHO none

\i test/setup.sql

-- test_role is set in test/deps.sql

SET LOCAL ROLE :use_role;
CREATE TEMP VIEW kinds AS
  SELECT kind, relkind
    FROM cat_tools.enum_range_srf('cat_tools.relation_kind') WITH ORDINALITY kind
      -- Use FULL OUTER to make sure we get errors if anything isn't in sync
      FULL OUTER JOIN cat_tools.enum_range_srf('cat_tools.relation_relkind') WITH ORDINALITY relkind
      USING(ordinality)
;

SELECT plan(
  1
  + 2 -- Simple is() tests
  + 4 -- no_use tests
  + 3 * (SELECT count(*)::int FROM kinds)
);

SELECT is(
  (SELECT count(*)::int FROM kinds)
  , 8
  , 'Verify count from kinds'
);

SELECT is(
  cat_tools.relation__kind('r')
  , 'table'::cat_tools.relation_kind
  , 'Simple sanity check of relation__kind()'
);
SELECT is(
  cat_tools.relation__relkind('table')
  , 'r'::cat_tools.relation_relkind
  , 'Simple sanity check of relation__relkind()'
);

SET LOCAL ROLE :no_use_role;
SELECT throws_ok(
  format( 'SELECT NULL::%I', typename )
  , '42704' -- undefined_object; not exactly correct, but close enough
  , NULL
  , 'Permission denied trying to use types'
)
  FROM (VALUES
    ('cat_tools.relation__relkind')
    , ('cat_tools.relation__kind')
  ) v(typename)
;
SELECT throws_ok(
  format( 'SELECT cat_tools.relation__%s( NULL::%I )', suffix, argtype )
  , '42501' -- insufficient_privilege
  , NULL
  , 'Permission denied trying to run functions'
)
  FROM (VALUES
    ('kind', 'text'::regtype)
    , ('relkind', 'text'::regtype)
  ) v(suffix, argtype)
;

SET LOCAL ROLE :use_role;

SELECT is(cat_tools.relation__relkind(kind)::text, relkind, format('SELECT cat_tools.relation_relkind(%L)', kind))
  FROM kinds
;

SELECT is(cat_tools.relation__kind(relkind)::text, kind, format('SELECT cat_tools.relation_kind(%L)', relkind))
  FROM kinds
;

SELECT is(cat_tools.relation__kind(relkind::"char")::text, kind, format('SELECT cat_tools.relation_kind(%L::"char")', relkind))
  FROM kinds
;

\i test/pgxntool/finish.sql

-- vi: expandtab ts=2 sw=2
