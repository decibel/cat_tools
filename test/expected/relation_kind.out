\set ECHO none
1..31
ok 1 - Verify count from kinds
ok 2 - Simple sanity check of relation__kind()
ok 3 - Simple sanity check of relation__relkind()
ok 4 - Permission denied trying to use types
ok 5 - Permission denied trying to use types
ok 6 - Permission denied trying to run functions
ok 7 - Permission denied trying to run functions
ok 8 - SELECT cat_tools.relation_relkind('table')
ok 9 - SELECT cat_tools.relation_relkind('index')
ok 10 - SELECT cat_tools.relation_relkind('sequence')
ok 11 - SELECT cat_tools.relation_relkind('toast table')
ok 12 - SELECT cat_tools.relation_relkind('view')
ok 13 - SELECT cat_tools.relation_relkind('materialized view')
ok 14 - SELECT cat_tools.relation_relkind('composite type')
ok 15 - SELECT cat_tools.relation_relkind('foreign table')
ok 16 - SELECT cat_tools.relation_kind('r')
ok 17 - SELECT cat_tools.relation_kind('i')
ok 18 - SELECT cat_tools.relation_kind('S')
ok 19 - SELECT cat_tools.relation_kind('t')
ok 20 - SELECT cat_tools.relation_kind('v')
ok 21 - SELECT cat_tools.relation_kind('c')
ok 22 - SELECT cat_tools.relation_kind('f')
ok 23 - SELECT cat_tools.relation_kind('m')
ok 24 - SELECT cat_tools.relation_kind('r'::"char")
ok 25 - SELECT cat_tools.relation_kind('i'::"char")
ok 26 - SELECT cat_tools.relation_kind('S'::"char")
ok 27 - SELECT cat_tools.relation_kind('t'::"char")
ok 28 - SELECT cat_tools.relation_kind('v'::"char")
ok 29 - SELECT cat_tools.relation_kind('c'::"char")
ok 30 - SELECT cat_tools.relation_kind('f'::"char")
ok 31 - SELECT cat_tools.relation_kind('m'::"char")
# TRANSACTION INTENTIONALLY LEFT OPEN!
